Move body writing logic into http-server

yawkat opened the pull request:
The main goal of this PR is to finally move the last pieces of the response encoding logic from RoutingInBoundHandler into a server-independent class (ResponseLifecycle). To accomplish this, I did a few refactors, mostly related to streaming responses.

Before this patch, streaming responses (e.g. Publisher<MyBean>) were handled by using the MessageBodyWriter to encode each item into a netty HttpContent (fully buffered) and concatenating them into a single netty ByteBody, with optional JSON separators (start/end bracket and commas).

The new approach is a bit different. The individual pieces are written using a new ResponseBodyWriter.writePiece method that, unlike the normal write method used before, does not have mutable response headers, which don't make sense for this scenario (the headers are already sent at the time of serialization). Additionally, it returns a ByteBody, so the piece can actually be streamed instead of fully buffered.

Next, the ByteBodys are concatenated using a new ConcatenatingSubscriber that streams each piece and forwards the data into a streaming ByteBody depending on runtime (BaseSharedBuffer implementation, for netty that's StreamingNettyByteBody). This class is the most intricate new code of this PR.

In order for servlet to use this ConcatenatingSubscriber, it needs a BaseSharedBuffer implementation, so I've moved the existing ReactiveByteBufferByteBody from the JDK client module to the HTTP module. For netty this is not used.

After these changes, I was able to move all the encoding methods to ResponseLifecycle with few changes. There are only a few netty-specific overrides in NettyResponseLifecycle: Netty-based ConcatenatingSubscribers, and handling  for the netty-specific StreamedHttpResponse.

To replace some of the netty-specific optimizations in RoutingInBoundHandler (such as the NettyResponseBodyWriterWrapper which has been removed), I've created a new ByteBodyFactory class that expands on the ByteBufferFactory concept with body factory methods. Unlike ByteBufferFactory, each method has a clear default implementation that works just fine and will be used as-is by servlet, but the class is also designed so that each method has a slightly more efficient netty implementation. ResponseBodyWriter has been changed to use ByteBodyFactory instead of ByteBufferFactory, making NettyResponseBodyWriterWrapper obsolete and also making many ResponseBodyWriters netty-independent (though I haven't moved them to http-server yet).

yawkat committed:
Move body writing logic into http-server
The main goal of this PR is to finally move the last pieces of the response encoding logic from RoutingInBoundHandler into a server-independent class (ResponseLifecycle). To accomplish this, I did a few refactors, mostly related to streaming responses.

Before this patch, streaming responses (e.g. Publisher<MyBean>) were handled by using the MessageBodyWriter to encode each item into a netty HttpContent (fully buffered) and concatenating them into a single netty ByteBody, with optional JSON separators (start/end bracket and commas).

The new approach is a bit different. The individual pieces are written using a new ResponseBodyWriter.writePiece method that, unlike the normal write method used before, does not have mutable response headers, which don't make sense for this scenario (the headers are already sent at the time of serialization). Additionally, it returns a ByteBody, so the piece can actually be streamed instead of fully buffered.

Next, the ByteBodys are concatenated using a new ConcatenatingSubscriber that streams each piece and forwards the data into a streaming ByteBody depending on runtime (BaseSharedBuffer implementation, for netty that's StreamingNettyByteBody). This class is the most intricate new code of this PR.

In order for servlet to use this ConcatenatingSubscriber, it needs a BaseSharedBuffer implementation, so I've moved the existing ReactiveByteBufferByteBody from the JDK client module to the HTTP module. For netty this is not used.

After these changes, I was able to move all the encoding methods to ResponseLifecycle with few changes. There are only a few netty-specific overrides in NettyResponseLifecycle: Netty-based ConcatenatingSubscribers, and handling  for the netty-specific StreamedHttpResponse.

To replace some of the netty-specific optimizations in RoutingInBoundHandler (such as the NettyResponseBodyWriterWrapper which has been removed), I've created a new ByteBodyFactory class that expands on the ByteBufferFactory concept with body factory methods. Unlike ByteBufferFactory, each method has a clear default implementation that works just fine and will be used as-is by servlet, but the class is also designed so that each method has a slightly more efficient netty implementation. ResponseBodyWriter has been changed to use ByteBodyFactory instead of ByteBufferFactory, making NettyResponseBodyWriterWrapper obsolete and also making many ResponseBodyWriters netty-independent (though I haven't moved them to http-server yet).

yawkat commented on http-client-jdk/src/main/java/io/micronaut/http/client/jdk/ByteBodySubscriber.java:
```diff
@@ -0,0 +1,135 @@
+ * @author Jonas Konrad
+ */
+@Internal
+final class ByteBodySubscriber implements HttpResponse.BodySubscriber<CloseableByteBody>, BufferConsumer.Upstream {
```
FYI: this has simply been moved from ReactiveByteBufferByteBody where it was an inner class previously, it's not new code.

graemerocher commented on http-netty/src/main/java/io/micronaut/http/netty/body/NettyByteBodyFactory.java:
```diff
@@ -0,0 +1,94 @@
+import java.nio.charset.StandardCharsets;
+
+@Internal
+public final class NettyByteBodyFactory extends ByteBodyFactory {
```
does this need to be public? If not even though it is internal please javadoc it

graemerocher commented on http-netty/src/main/java/io/micronaut/http/netty/body/NettyWritableBodyWriter.java:
```diff
@@ -74,19 +77,24 @@ public boolean isBlocking() {
 
     @Override
-    public ByteBodyHttpResponse<?> write(ByteBufferFactory<?, ?> bufferFactory, HttpRequest<?> request, MutableHttpResponse<Writable> outgoingResponse, Argument<Writable> type, MediaType mediaType, Writable object) throws CodecException {
+    public ByteBodyHttpResponse<?> write(@NonNull ByteBodyFactory bodyFactory, HttpRequest<?> request, MutableHttpResponse<Writable> outgoingResponse, Argument<Writable> type, MediaType mediaType, Writable object) throws CodecException {
```
break the arguments on multiple lines to make the method easier to read

graemerocher commented on http-netty/src/main/java/io/micronaut/http/netty/body/NettyWritableBodyWriter.java:
```diff
@@ -74,19 +77,24 @@ public boolean isBlocking() {
+    }
+
+    @Override
+    public CloseableByteBody writePiece(@NonNull ByteBodyFactory bodyFactory, @NonNull HttpRequest<?> request, @NonNull HttpResponse<?> response, @NonNull Argument<Writable> type, @NonNull MediaType mediaType, Writable object) {
```
break the arguments on multiple lines to make the method easier to read

graemerocher commented on http-server-netty/src/main/java/io/micronaut/http/server/netty/NettyResponseLifecycle.java:
```diff
@@ -0,0 +1,186 @@
+    private final NettyHttpRequest<?> request;
+
+    public NettyResponseLifecycle(RoutingInBoundHandler routingInBoundHandler, NettyHttpRequest<?> request) {
+        super(routingInBoundHandler.routeExecutor, routingInBoundHandler.messageBodyHandlerRegistry, routingInBoundHandler.conversionService, new NettyByteBodyFactory(request.getChannelHandlerContext().channel()));
```
break the arguments on multiple lines to make the method easier to read

graemerocher commented on http-server-netty/src/main/java/io/micronaut/http/server/netty/NettyResponseLifecycle.java:
```diff
@@ -0,0 +1,186 @@
+        return JsonNettyConcatenatingSubscriber.concatenateJson(eventLoop(), items);
+    }
+
+    private static class NettyConcatenatingSubscriber extends ConcatenatingSubscriber implements ByteBufConsumer {
```
declare `final` if possible

yawkat replied:
It's extended by the JsonNettyConcatenatingSubscriber which adds the json array stuff (`[`, `,`, `]`)

graemerocher commented on http-server-netty/src/main/java/io/micronaut/http/server/netty/NettyResponseLifecycle.java:
```diff
@@ -0,0 +1,186 @@
+    }
+
+    private static class NettyConcatenatingSubscriber extends ConcatenatingSubscriber implements ByteBufConsumer {
+        final StreamingNettyByteBody.SharedBuffer sharedBuffer;
```
does this need to be package protected?

yawkat replied:
yes, the subclass accesses it

graemerocher commented on http-server-netty/src/main/java/io/micronaut/http/server/netty/NettyResponseLifecycle.java:
```diff
@@ -0,0 +1,186 @@
+            sharedBuffer = new StreamingNettyByteBody.SharedBuffer(eventLoop, BodySizeLimits.UNLIMITED, this);
+        }
+
+        public static CloseableByteBody concatenate(EventLoop eventLoop, Publisher<ByteBody> publisher) {
```
does this need to be public?

graemerocher commented on http-server-netty/src/main/java/io/micronaut/http/server/netty/body/FileBodyWriter.java:
```diff
@@ -49,10 +52,15 @@ public FileBodyWriter(SystemFileBodyWriter systemFileBodyWriter) {
 
     @Override
-    public ByteBodyHttpResponse<?> write(ByteBufferFactory<?, ?> bufferFactory, HttpRequest<?> request, MutableHttpResponse<File> outgoingResponse, Argument<File> type, MediaType mediaType, File object) throws CodecException {
+    public ByteBodyHttpResponse<?> write(@NonNull ByteBodyFactory bodyFactory, HttpRequest<?> request, MutableHttpResponse<File> outgoingResponse, Argument<File> type, MediaType mediaType, File object) throws CodecException {
```
break the arguments on multiple lines to make the method easier to read

graemerocher commented on http-server-netty/src/main/java/io/micronaut/http/server/netty/body/FileBodyWriter.java:
```diff
@@ -49,10 +52,15 @@ public FileBodyWriter(SystemFileBodyWriter systemFileBodyWriter) {
+    }
+
+    @Override
+    public CloseableByteBody writePiece(@NonNull ByteBodyFactory bodyFactory, @NonNull HttpRequest<?> request, @NonNull HttpResponse<?> response, @NonNull Argument<File> type, @NonNull MediaType mediaType, File object) {
```
break the arguments on multiple lines to make the method easier to read

graemerocher commented on http-server-netty/src/main/java/io/micronaut/http/server/netty/body/InputStreamBodyWriter.java:
```diff
@@ -58,9 +60,14 @@ public final class InputStreamBodyWriter extends AbstractFileBodyWriter implemen
 
     @Override
-    public ByteBodyHttpResponse<?> write(ByteBufferFactory<?, ?> bufferFactory, HttpRequest<?> request, MutableHttpResponse<InputStream> outgoingResponse, Argument<InputStream> type, MediaType mediaType, InputStream object) throws CodecException {
+    public ByteBodyHttpResponse<?> write(@NonNull ByteBodyFactory bodyFactory, HttpRequest<?> request, MutableHttpResponse<InputStream> outgoingResponse, Argument<InputStream> type, MediaType mediaType, InputStream object) throws CodecException {
```
break the arguments on multiple lines to make the method easier to read

graemerocher commented on http-server-netty/src/main/java/io/micronaut/http/server/netty/body/InputStreamBodyWriter.java:
```diff
@@ -58,9 +60,14 @@ public final class InputStreamBodyWriter extends AbstractFileBodyWriter implemen
+    }
+
+    @Override
+    public CloseableByteBody writePiece(@NonNull ByteBodyFactory bodyFactory, @NonNull HttpRequest<?> request, @NonNull HttpResponse<?> response, @NonNull Argument<InputStream> type, @NonNull MediaType mediaType, InputStream object) {
```
break the arguments on multiple lines to make the method easier to read

graemerocher commented on http-server-netty/src/main/java/io/micronaut/http/server/netty/body/StreamFileBodyWriter.java:
```diff
@@ -58,16 +60,21 @@ public final class StreamFileBodyWriter extends AbstractFileBodyWriter implement
 
     @Override
-    public ByteBodyHttpResponse<?> write(ByteBufferFactory<?, ?> bufferFactory, HttpRequest<?> request, MutableHttpResponse<StreamedFile> outgoingResponse, Argument<StreamedFile> type, MediaType mediaType, StreamedFile object) throws CodecException {
+    public ByteBodyHttpResponse<?> write(@NonNull ByteBodyFactory bodyFactory, HttpRequest<?> request, MutableHttpResponse<StreamedFile> outgoingResponse, Argument<StreamedFile> type, MediaType mediaType, StreamedFile object) throws CodecException {
```
break the arguments on multiple lines to make the method easier to read

graemerocher commented on http-server-netty/src/main/java/io/micronaut/http/server/netty/body/StreamFileBodyWriter.java:
```diff
@@ -58,16 +60,21 @@ public final class StreamFileBodyWriter extends AbstractFileBodyWriter implement
     }
 
+    @Override
+    public CloseableByteBody writePiece(@NonNull ByteBodyFactory bodyFactory, @NonNull HttpRequest<?> request, @NonNull HttpResponse<?> response, @NonNull Argument<StreamedFile> type, @NonNull MediaType mediaType, StreamedFile object) {
```
break the arguments on multiple lines to make the method easier to read

graemerocher commented on http-server-netty/src/main/java/io/micronaut/http/server/netty/body/SystemFileBodyWriter.java:
```diff
@@ -132,10 +133,26 @@ public ByteBodyHttpResponse<?> write(HttpRequest<?> request, MutableHttpResponse
     }
 
+    @Override
+    public CloseableByteBody writePiece(@NonNull ByteBodyFactory bodyFactory, @NonNull HttpRequest<?> request, @NonNull HttpResponse<?> response, @NonNull Argument<SystemFile> type, @NonNull MediaType mediaType, SystemFile object) {
```
break the arguments on multiple lines to make the method easier to read

graemerocher commented on http-server-netty/src/main/java/io/micronaut/http/server/netty/body/SystemFileBodyWriter.java:
```diff
@@ -132,10 +133,26 @@ public ByteBodyHttpResponse<?> write(HttpRequest<?> request, MutableHttpResponse
+        try {
+            is = new FileInputStream(object.getFile());
+        } catch (FileNotFoundException e) {
+            throw new MessageBodyException("Could not find file", e);
```
should the file name be included?

yawkat replied:
I think I'd rather not in case it is user-controlled in some way

graemerocher commented on http-server/src/main/java/io/micronaut/http/server/ResponseLifecycle.java:
```diff
@@ -0,0 +1,345 @@
+    private final ConversionService conversionService;
+    private final ByteBodyFactory byteBodyFactory;
+
+    public ResponseLifecycle(RouteExecutor routeExecutor, MessageBodyHandlerRegistry messageBodyHandlerRegistry, ConversionService conversionService, ByteBodyFactory byteBodyFactory) {
```
break the arguments on multiple lines to make the method easier to read

yawkat committed:
changes necessary for servlet

dstepanov commented on http/src/main/java/io/micronaut/http/body/ResponseBodyWriter.java:
```diff
@@ -40,24 +40,46 @@ public interface ResponseBodyWriter<T> extends MessageBodyWriter<T> {
         T object) throws CodecException;
 
+    /**
+     * Write a <i>piece</i> of a larger response, e.g. when writing a Publisher or a part of a
```
What exactly is a `piece` of the response? A piece of the body? Should it be `writebodyPart`?  What is the relation to the returned closeable body? How is supposed to close it?

yawkat replied:
A "piece" can be a part (as in multipart), but right now it's only a piece of a streamed response (e.g. when you have a streamed json array). I called it piece instead of part to avoid giving the impression that this is only for multipart, though I plan to use this for multipart as well.

The returned body is Closeable because ownership transfers to the caller. The caller has to either close the body or use it (e.g. concatenating for the json array use case).

graemerocher replied:
is `Chunk` a better name? That is what Netty refers to it is it not?

yawkat replied:
"chunk" is already used in the context `transfer-encoding: chunked`, so it is not really appropriate either

dstepanov commented on http-server/src/main/java/io/micronaut/http/server/HttpServerConfiguration.java:
```diff
@@ -1045,4 +1051,116 @@ public void setHeader(boolean header) {
+     * @since 3.1.0
+     */
+    @ConfigurationProperties("responses.file")
+    public static class FileTypeHandlerConfiguration {
```
Is is a new class?

yawkat replied:
I'm just moving the `micronaut.server.netty.responses.file` config to `micronaut.server.responses.file`, not sure if this is the best way

yawkat committed:
properly handle pre-acknowledged bytes

graemerocher commented on http/src/main/java/io/micronaut/http/body/ByteBodyFactory.java:
```diff
@@ -0,0 +1,161 @@
+     * @param byteBufferFactory The buffer factory
+     */
+    @Internal
+    protected ByteBodyFactory(@NonNull ByteBufferFactory<?, ?> byteBufferFactory) {
```
is this designed for extension?

yawkat replied:
Yes, by NettyByteBodyFactory

graemerocher commented on http/src/main/java/io/micronaut/http/body/ByteBufferBodyAdapter.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    @NonNull
+    static ReactiveByteBufferByteBody adapt(Publisher<ByteBuffer> source) {
```
javadoc

graemerocher commented on http/src/main/java/io/micronaut/http/body/ByteBufferBodyAdapter.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    @NonNull
+    static ReactiveByteBufferByteBody adapt(Publisher<ByteBuffer> publisher, @Nullable HttpHeaders headersForLength, @Nullable Runnable onDiscard) {
```
javadoc

graemerocher commented on http/src/main/java/io/micronaut/http/body/ConcatenatingSubscriber.java:
```diff
@@ -0,0 +1,322 @@
+ * @author Jonas Konrad
+ */
+@Internal
+public abstract class ConcatenatingSubscriber implements BufferConsumer.Upstream, CoreSubscriber<ByteBody>, BufferConsumer {
```
can this be sealed?

yawkat replied:
No, it's extended by netty

yawkat committed:
CR

yawkat committed:
Add note on NettyByteBodyFactory constructor

yawkat committed:
Merge branch '4.8.x' into abstract-body-writing

# Conflicts:
#	http-server/src/main/java/io/micronaut/http/server/HttpServerConfiguration.java

graemerocher commented on core/src/main/java/io/micronaut/core/util/functional/ThrowingConsumer.java:
```diff
@@ -0,0 +1,35 @@
+     * @param t The value
+     * @throws E The generic exception
+     */
+    void accept(T t) throws E;
```
add nullability annotation to the parameter

graemerocher commented on http-server-netty/src/main/java/io/micronaut/http/server/netty/configuration/NettyHttpServerConfiguration.java:
```diff
@@ -1196,7 +1196,9 @@ public static class Parent extends EventLoopConfig {
      * @since 3.1.0
+     * @deprecated Replaced by {@link HttpServerConfiguration.FileTypeHandlerConfiguration}
      */
+    @Deprecated
```
add `@since` and forremoval

graemerocher commented on http-server-netty/src/main/java/io/micronaut/http/server/netty/configuration/NettyHttpServerConfiguration.java:
```diff
@@ -1268,7 +1262,10 @@ public void setCacheControl(CacheControlConfiguration cacheControl) {
+         * 
+         * @deprecated Replaced by {@link HttpServerConfiguration.FileTypeHandlerConfiguration.CacheControlConfiguration}
          */
+        @Deprecated
```
add `@since` and forRemoval

yawkat committed:
fix backpressure in ConcatenatingSubscriber for immediate bodies

yawkat committed:
review comments

yawkat committed:
fix reentrancy in LazySendingSubscriber

yawkat committed:
Fix request leak for null response
When a connection is closed, the response processing ExecutionFlow may return a `null` response. This patch replaces that with a mock response so that the responseWritten logic is called and the request lifecycle finishes properly.

Found by fuzzing.

yawkat commented:
I've fixed another bug found by the fuzzer in https://github.com/micronaut-projects/micronaut-core/pull/11342/commits/8d5976d5f721340c1ea207fab80bcd0bfaddfdd3 . I made the change in this PR because they would conflict otherwise.
